datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java

```diff
@@ -406,16 +390,74 @@ private void reportMetrics(DatastreamRecordMetadata metadata, long eventsSourceT
     // Report the time it took to just send the events to destination
     if (eventsSendTimestamp > 0) {
       long sendLatency = System.currentTimeMillis() - eventsSendTimestamp;
-      _dynamicMetricsManager.createOrUpdateHistogram(MODULE, topicOrDatastreamName, EVENTS_SEND_LATENCY_MS_STRING,
-          sendLatency);
-      _dynamicMetricsManager.createOrUpdateHistogram(MODULE, AGGREGATE, EVENTS_SEND_LATENCY_MS_STRING, sendLatency);
-      _dynamicMetricsManager.createOrUpdateHistogram(MODULE, _datastreamTask.getConnectorType(),
-          EVENTS_SEND_LATENCY_MS_STRING, sendLatency);
+      reportSendLatencyMetrics(metadata, sendLatency, EVENTS_SEND_LATENCY_MS_STRING);
+    }
+    _dynamicMetricsManager.createOrUpdateMeter(MODULE, AGGREGATE, EVENT_PRODUCE_RATE, 1);
+    _dynamicMetricsManager.createOrUpdateMeter(MODULE, _datastreamTask.getConnectorType(), EVENT_PRODUCE_RATE, 1);
+  }
+
+  /**
+   * Only for the throughput violating topics!
+   * <br>
+   * <br>
+   * Report metrics on every send callback from the transport provider. Because this can be invoked multiple times
+   * per DatastreamProducerRecord (i.e. by the number of events within the record), only increment all metrics by 1
+   * to avoid overcounting.
+   */
+  private void reportMetricsForThroughputViolatingTopics(DatastreamRecordMetadata metadata, long eventsSourceTimestamp,
+      long eventsSendTimestamp) {
+    // Treat all events within this record equally (assume same timestamp)
+    if (eventsSourceTimestamp > 0) {
+      // Report availability metrics
+      reportEventLatencyMetrics(metadata, eventsSourceTimestamp, THROUGHPUT_VIOLATING_EVENTS_LATENCY_MS_STRING);
+      _dynamicMetricsManager.createOrUpdateCounter(MODULE, AGGREGATE, TOTAL_EVENTS_PRODUCED, 1);
+      _dynamicMetricsManager.createOrUpdateCounter(MODULE, _datastreamTask.getConnectorType(), TOTAL_EVENTS_PRODUCED,
```

Comment:
This requires more precise documentation explaining X events from the throughput violation topic and Y events from the regular topic and total records X+Y.